### PR TITLE
net: fix TcpSocket bugs introduced in #4270

### PR DESCRIPTION
**Pull request by taiki-e**
This is an attempt to fix bugs of TcpSocket that I accidentally introduced in #4270.

This should pass tokio's test suite, but tokio has very few tests for TcpSocket (this patch adds a few simple assertions), so we need to run hyper's test suite to see if it works correctly.

~~This patch is still incomplete, and the following twoone hyper's tests are currently failing.~~

UPD: Timeout in hyper/benches/connect.rs seems to be unrelated to #4270 as it can be reproduced in tokio v1.15.0.
UPD2: Double panic in hyper/tests/integration.rs is due to #4377

~~I have very little bandwidth on the weekdays, so if anyone knows how to fix these issues, feel free to copy this patch and open a new PR.~~

### CODE REVIEW

**Comment by taiki-e** [comment]
Oh, timeout in hyper/benches/connect.rs doesn't seem to happen on linux?: https://github.com/tokio-rs/tokio/runs/4775733526?check_suite_focus=true

File: tokio/src/net/tcp/socket.rs [hunk 1]
```diff
@@ -125,6 +125,7 @@ impl TcpSocket {
             socket2::Type::STREAM,
             Some(socket2::Protocol::TCP),
         )?;
+        inner.set_nonblocking(true)?;
```
Comment (carllerche) [inline tokio/src/net/tcp/socket.rs:128]: There *should* be a way to do this as part of the `Socket::new` initialization. If there isn't, we need to fix that.

Comment (Thomasdezeeuw) [reply tokio/src/net/tcp/socket.rs:128]: Socket2 supports [`Type::nonblocking`](https://docs.rs/socket2/latest/socket2/struct.Type.html#method.nonblocking) for this.

File: tokio/src/net/tcp/socket.rs [hunk 2]
```diff
@@ -534,7 +536,15 @@ impl TcpSocket {
     /// }
     /// ```
     pub async fn connect(self, addr: SocketAddr) -> io::Result<TcpStream> {
-        self.inner.connect(&addr.into())?;
+        match self.inner.connect(&addr.into()) {
+            // https://github.com/tokio-rs/mio/blob/d400ddfc97212b4a2844d741b095dab2c6d15543/src/sys/unix/tcp.rs#L31
+            #[cfg(unix)]
+            Err(e) if e.raw_os_error() != Some(libc::EINPROGRESS) => return Err(e),
```
Comment (Thomasdezeeuw) [inline tokio/src/net/tcp/socket.rs:542]: @carllerche @seanmonstar I think this should fix the issue.

File: tokio/src/net/tcp/socket.rs (hunk 1 above)
Comment (taiki-e) [reply tokio/src/net/tcp/socket.rs]: addressed in 08240cc.

**Review by Thomasdezeeuw — APPROVED** [review]
Change to `Socket::new` and `connect` LGTM, didn't check the tests.

**Comment by taiki-e** [comment]
The double panic from hyper's test suite also seems to be unrelated to this problem.

I can confirm the double panic on #4391.

**Comment by taiki-e** [comment]
Ok, reverting https://github.com/tokio-rs/tokio/commit/cc8ad367a0e5d8536f8be58fe560bfdea1a976a5 solves the remaining problems.

**Comment by Noah-Kennedy** [comment]
I can try this tonight, I have access to MacOS and Windows machines for development.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -355,3 +356,36 @@ jobs:
       - name: Verify that Tokio version is up to date in README
         working-directory: tokio
         run: grep -q "$(sed '/^version = /!d' Cargo.toml | head -n1)" README.md
+
+  test-hyper:
+    name: Test hyper
+    runs-on: ubuntu-latest
```
Comment (carllerche) [inline .github/workflows/ci.yml:362]: Could we run hyper tests on other platforms as well?

Comment (taiki-e) [reply .github/workflows/ci.yml:362]: #4393 (CI part of this PR) contains changes for that.

**Comment by carllerche** [comment]
I'm going to close this PR since the changes being fixed were reverted and the CI improvements were merged separately. We need to figure out how we want to move forward w/ mio 0.8 still.

**Comment by Thomasdezeeuw** [comment]
> I'm going to close this PR since the changes being fixed were reverted and the CI improvements were merged separately. We need to figure out how we want to move forward w/ mio 0.8 still.

This is a rather surprising reaction, especially considering the bug wasn't even in Mio. Really the only problem was the missing check for `EINPROGRESS` which this pr would fix.

**Pull request closed without merging**